torch/_inductor/select_algorithm.py

```diff
@@ -1125,215 +1125,237 @@ def __call__(
         precompilation_timeout_seconds: int = 60 * 60,
         return_multi_template=False,
     ):
-        from .codegen.cuda.cuda_kernel import CUDATemplateCaller
-
-        # Templates selected with input_gen_fns require specific input data to avoid IMA
-        # Passing custom input gen fns to benchmark_fusion NYI, so skip deferred template selection
-        # TODO(jgong5): support multi-template on CPU
-        if input_gen_fns is not None or layout.device.type == "cpu":
-            return_multi_template = False
-
-        # TODO - assert that we have not mutating kernels here
-
-        # TODO(nmacchioni): remove once CI tests are fixed
-        choices = [choice for choice in choices if choice is not None]
-
-        if mm_file_name := get_mm_log_filename():
-            M, K = input_nodes[-2].get_size()[:2]
-            N = input_nodes[-1].get_size()[-1]
-            append_to_log(mm_file_name, {"invoke": str((M, K, N))})
+        assert config.search_autotune_cache or config.max_autotune_gemm or config.max_autotune
 
         if len(choices) == 0:
             raise NoValidChoicesError(
                 "No choices to select, please consider adding ATEN into max_autotune_gemm_backends "
                 "config (defined in torch/_inductor/config.py) to allow at least one choice. "
             )
-        log.debug("Max autotune selects from %s choices.", str(len(choices)))
-
-        if len(choices) == 1:
+        elif len(choices) == 1:
+            from .codegen.cuda.cuda_kernel import CUDATemplateCaller
             if not isinstance(choices[0], CUDATemplateCaller):
-                # CUDATemplateCaller still needs to go through autotuning process to retrieve workspace size.
+                # CUDATemplateCaller still needs to go through autotuning process to retrieve workspace size
                 return choices[0].output_node()
 
-        @functools.lru_cache(None)
-        def make_benchmark_fn():
-            return self.make_benchmark_fn(choices, input_nodes, layout, input_gen_fns)
-
         inputs_key = repr([self.key_of(x) for x in input_nodes])
 
-        def precompile(choices) -> Callable[[], None]:
-            def no_op(*args, **kwargs):
-                return
+        precompile = functools.partial(self.precompile, name, inputs_key)
+        
+        if config.search_autotune_cache:
```

Comment:
Maybe we should always search the cache first, even of config.search_autotune_cache is not explicitly set? I feel like config.max_autotune_gemm and config.max_autotune should auto-opt into this feature (especially since it is done anyways in the actual autotuning stage)